### PR TITLE
3323 combobox tanstack virtual

### LINKED ISSUE #3269: [COMBOBOX] Items in a long list are not displayed
### Provide a general summary of the issue here

If `sl-combobox` shows a long list, items near the end of the list are not displayed

2.  type option 50 to select this item from the select
3. select the item
4. leave the comboxbox field
5. enter the field and the list is empty

Image

Please use the story below to reproduce this issue

```typescript
export const FilterResults: Story = {
  render: () => {
    return html`
     <sl-combobox
        filter-results
        .options=${Array.from({ length: 100 }, (_, i) => `Option ${i + 1}`)}
      >
      </sl-combobox>

      <sl-text-field></sl-text-field>    
      `
  }
};
```

### 👤 Your name

Frans de Boer

### 🧢 Your Product/Team

Magister

### LINKED ISSUE #3323: [Combobox] List of options is not visible for stories Single Combobox Virtual List and Virtual List with Options
### 😯 Current Behavior

This happens in Safari on macOS and Firefox on both Win11 and macOS. 
List of options appears empty for stories Single Combobox Virtual List and Virtual List with Options.

Safari:

https://github.com/user-attachments/assets/1352b945-ba37-461c-969f-04e354a5e5b1

Firefox:

https://github.com/user-attachments/assets/b4ac5558-0e66-4ac6-9c34-0922acacb721

### 🤔 Expected Behavior/possible solution?

https://github.com/user-attachments/assets/81a67fae-956f-49f3-a317-bdf36eac4333

### 🖥️ Steps to Reproduce

1. In Safari or Firefox go to https://storybook.sanomalearning.design/?path=/story/form-combobox-single--virtual-list
2. Click on option that is displayed on inside combobox. 
3. Check if listbox that appeared has visible options inside of it.

### Were you able to reproduce this in our storybook?

Yes

### If yes, provide the storybook URL

https://storybook.sanomalearning.design/?path=/story/form-combobox-single--virtual-list

### If no, provide a code reproduction

_No response_

### ⚖️ Impact

This blocks our release or stops us from using this component

### Framework

Other

### What browsers are you seeing the problem on?

Firefox, Safari

### If other, please specify.

_No response_

### What operating system are you using?

macOS and Win11

### Component version

v0.1.8

### Theme version

_No response_

### 👤 Your name

Kacper

### 🧢 Your product/team

SLDS

### 🎨 Your Theme(s)

Sanoma Learning

### PR DESCRIPTION
**Pull request by Diaan**
This pull request introduces major improvements to the `@sl-design-system/listbox`, `@sl-design-system/combobox`, and `@sl-design-system/virtual-list` packages, focusing on enhanced performance and user experience when handling large option lists. The most significant change is the migration of the virtual list implementation to use TanStack Virtual, resulting in smoother scrolling and more reliable virtualization. Additionally, the combobox and listbox components have been updated to better handle scrolling, selection, and constraints, with improved CSS handling and more robust test coverage.

**Performance and Virtualization Improvements:**
- Refactored `@sl-design-system/virtual-list` to use TanStack Virtual for core virtualization logic, leading to better performance and smoother scrolling. Added new features such as the `scrollMargin` property and `requestLayout()` method.
- Updated `@sl-design-system/listbox` to integrate the new virtual-list component, significantly improving performance and scrolling behavior for large lists. Added support for CSS max-height constraints.
- Improved scroll-to-item behavior and virtual list handling in `@sl-design-system/combobox`, depending on the updated listbox and virtual-list packages.

**Component and UX Enhancements:**
- Enhanced keyboard navigation and option selection in `combobox.ts` by making scroll behavior smoother and more consistent across browsers. Improved handling of current item updates and ensured correct scrolling for both virtual and non-virtual lists. [[1]](diffhunk://#diff-8f9a70134090f1e8edb4e27b08e8b9849ea6c405ed1572a298eeb85a3e3fac7bL771-R771) [[2]](diffhunk://#diff-8f9a70134090f1e8edb4e27b08e8b9849ea6c405ed1572a298eeb85a3e3fac7bL905-R905) [[3]](diffhunk://#diff-8f9a70134090f1e8edb4e27b08e8b9849ea6c405ed1572a298eeb85a3e3fac7bL1394-R1430)
- Updated CSS for `combobox` and `listbox` to better handle flex layouts and provide a max-block-size fallback for unconstrained virtual lists, with the ability for consumers to override via inline styles or CSS. [[1]](diffhunk://#diff-8d58c244e0a7dcafc370ba70378f95ad96da1c1341dc9d912a03d51ac28c2041L24-R25) [[2]](diffhunk://#diff-56c644c1149a7baede4daa98d0040af73f97c6927bdf597bbee1a104f3db1259R13-R16)

**Testing and Storybook Updates:**
- Added and updated tests for virtual list rendering, scroll behavior, and CSS constraint handling in both combobox and listbox. Ensured robust coverage for edge cases and fallback behaviors. [[1]](diffhunk://#diff-35008062a8c72c8035b697e5604e597bb52badfa6e2641d8d2e681f866495904R1660-R1714) [[2]](diffhunk://#diff-2cb85c9d2a2a2e6dc6ed44627dc8c0ae0106536efd6781555a72036ae300f730L107-R228)
- Updated Storybook stories for listbox to support new scrolling behaviors and icons, improving documentation and demo capabilities. [[1]](diffhunk://#diff-7e1fa65739f25d1df17459a6861a6c53e8e861800cdd08420c84869f5335e9abR1-R17) [[2]](diffhunk://#diff-7e1fa65739f25d1df17459a6861a6c53e8e861800cdd08420c84869f5335e9abR29-R39)

**Dependency and Internal Refactoring:**
- Updated dependencies in `listbox/package.json` to use the new virtual-list package and removed the previous `@lit-labs/virtualizer` dependency.

These changes collectively modernize the listbox and combobox components, making them more scalable and user-friendly when working with large datasets.

### CODE REVIEW

**Review by a11ymiko — COMMENTED** [review]
On Chrome when navigating to previous items with up arrow key user cant see focused element. They can still use this invisible element with Enter.

https://github.com/user-attachments/assets/82b3773b-a760-46f8-99d2-d27355f2d2e3

**Review by a11ymiko — COMMENTED** [review]
When user goes to the bottom of options end from the last option goes to smaller numbers with arrow keys focus indicator is being displayed on each of options at the same time. 
It starts to behave normally once it reaches 'Option 9991' but when user go back to 'Option 9992' and next ones focus indicator is displayed on multiple options once again: 

https://github.com/user-attachments/assets/86d186e7-6469-4458-b20b-c62ad404ec89

**Review by a11ymiko — COMMENTED** [review]
User can go with arrow keys from the last option to the first one and then continue to go to 'Option 2', 'Option 3' etc. but the list still display 'Option 9991'-'Option 10000'. 
User can select focused option, but they don't know what the option is.

https://github.com/user-attachments/assets/cdf38028-71ca-4811-a459-76bdfdd25f92

**Review by a11ymiko — COMMENTED** [review]
Mouse user cannot scroll up. The list only allows to be scrolled down. Even when some options were already loaded and displayed moment ago when they are not currently displayed then user cannot scroll up back to them

https://github.com/user-attachments/assets/3c1af9a6-ffb4-41df-9440-fb47b0f097d9

File: packages/components/virtual-list/package.json [hunk 1]
```diff
@@ -42,11 +42,11 @@
     "@sl-design-system/shared": "^0.12.0"
   },
   "devDependencies": {
-    "@tanstack/virtual-core": "^3.13.23",
+    "@tanstack/virtual-core": "3.13.12",
     "lit": "^3.3.2"
   },
   "peerDependencies": {
-    "@tanstack/virtual-core": "^3.13.12",
+    "@tanstack/virtual-core": "3.13.12",
```
[the remark this answers is not in the document]
Comment (Diaan) [reply packages/components/virtual-list/package.json:49]: This is deliberate; there are bugs in newer versions.

File: packages/components/virtual-list/package.json [hunk 2]
```diff
@@ -42,11 +42,11 @@
...
     "lit": "^3.1.4"
```
[the remark this answers is not in the document]
Comment (Diaan) [reply packages/components/virtual-list/package.json:50]: this is deliberate, there are bugs in the newer version

**Review by a11ymiko — COMMENTED** [review]
In Virtual List With Groups story button "Scroll to 5000" takes user to "Option 4950" instead.

https://github.com/user-attachments/assets/0ef863a2-bec2-4269-aeb0-f938a558f77a

**Review by a11ymiko — APPROVED** [review]

**Review by michal-sanoma — APPROVED** [review]
Not sure if we need this note 😃

Image

File: packages/components/listbox/src/listbox.scss [hunk 3]
```diff
@@ -10,6 +10,10 @@
   scrollbar-width: thin;
 }
 
+:host([data-virtual-unconstrained]) {
+  max-block-size: 20rem;
```
Comment (anna-lach) [inline packages/components/listbox/src/listbox.scss:17]: Do we want to keep that value this way?

Comment (anna-lach) [reply packages/components/listbox/src/listbox.scss:17]: Maybe it's worth to leave a comment what's the purpose and why such value.

File: packages/components/listbox/src/listbox.ts [hunk 4]
```diff
@@ -146,41 +155,101 @@ export class Listbox<T = any, U = T> extends ScopedElementsMixin(LitElement) {
...
+    // Set attribute if no constraints found
+    if (
+      this.items &&
+      this.items.length > 0 &&
+      !hasInlineHeightConstraint &&
+      !hasComputedHeightConstraint
+    ) {
+      this.setAttribute('data-virtual-unconstrained', '');
+    }
+    // else: attribute already removed above
```
Comment (anna-lach) [inline packages/components/listbox/src/listbox.ts]: Is that comment necessary?

File: packages/components/virtual-list/src/virtual-list.stories.ts [hunk 5]
```diff
@@ -109,11 +141,22 @@ export default {
...
       <sl-button-bar>
-        <sl-button @click=${() => scrollTo(0)}>Scroll to top</sl-button>
-        <sl-button @click=${() => scrollTo(items.length / 2)}
-          >Scroll to ${items.length / 2}</sl-button
+        Scroll:
+        <sl-button @click=${() => scrollTo(0)} aria-label="Scroll to top"
+          ><sl-icon name="far-arrow-up-to-line"></sl-icon
+        ></sl-button>
+        <sl-button @click=${() => scrollTo(scrollToPosition - 1)} aria-label="Scroll up one"
+          ><sl-icon name="far-arrow-up"></sl-icon
```
Comment (anna-lach) [inline packages/components/virtual-list/src/virtual-list.stories.ts:149]: When I go to the end of the list and then try to go to the top, it looks like it's not always working.

https://github.com/user-attachments/assets/3dabc561-9c3e-4b07-95bf-91508ea827d4

**Review by anna-lach — APPROVED** [review]
